src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WinHttpWebSocket.cs

```diff
@@ -441,6 +448,12 @@ private Task<bool> InternalReceiveAsync(ArraySegment<byte> buffer)
                 Abort();
             }
 
+            // If the task has completed, there's no need to force continuations to run async
+            if (_operation.TcsReceive.Task.Status == TaskStatus.RanToCompletion)
+            {
+                return Task.FromResult(_operation.TcsReceive.Task.GetAwaiter().GetResult());
+            }
```

Comment:
I think the misunderstanding is around how WinHttp handles callbacks. There is a global callback function registered with WinHttp which has a parameter saying which event has completed. When you call for example WinHttpWebSocketReceive, if there's data available (sync completion), it calls the callback from that method on the same thread. If there isn't data available (async completion), then the method returns and will call the callback on a WinHttp owned thread when there is data available. In the callback, the managed code checks the callback reason, see's it is a read completion and calls a method to handle the read completion. In that method, it fetches the result data and then sets the result on the TCS. If this is an async continuation case, WCF code has registered a continuation on the returned Task (regular await scenario) which then gets run when the TCS has a result set. At this point, the await continuation code is running on the WinHttp owned callback thread. WCF will make async calls whenever it can which is the case on an initial ReceiveAsync call when waiting for an incoming WCF message, but due to serializer/XmlReader constraints, some parts of our code can't always call code asynchronously. So from the WinHttp owned thread, we call ReceiveAsync again, but now we're in some WCF code which is on a synchronous call stack and WCF must Wait() the returned Task. The call to ReceiveAsync goes into WinHttpWebSocketReceive which won't call the callback as it doesn't support being called reentrant and returns the uncompleted Task. WCF then calls Wait() on the Task resulting in blocking the thread, but WinHttp won't call the callback (and so won't complete the Task) while we are still in control of it's callback thread. So we end up deadlocked.  
The intention of my PR is to elicit the behavior of completing on the same thread if the WinHttp call completed synchronously, or forcing an async completion (get off of the WinHttp owned thread) when completing async.